lib/matplotlib/_constrained_layout.py

```diff
@@ -83,181 +74,357 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
+            maxsubl = np.max(
+                lg1.margin_vals['left'][ss1.colspan[1:]] +
+                lg1.margin_vals['leftcb'][ss1.colspan[1:]]
+            )
+            maxsubr = np.max(
+                lg1.margin_vals['right'][ss1.colspan[:-1]] +
+                lg1.margin_vals['rightcb'][ss1.colspan[:-1]]
+            )
+            for ax2 in axs:
+                ss2 = ax2.get_subplotspec()
+                lg2 = ss2.get_gridspec()._layoutgrid
+                if lg2 is not None:
+                    nc = len(ss2.colspan)
+                    if nc > 1:
+                        maxsubl2 = np.max(
+                            lg2.margin_vals['left'][ss2.colspan[1:]] +
+                            lg2.margin_vals['leftcb'][ss2.colspan[1:]])
+                        if maxsubl2 > maxsubl:
+                            maxsubl = maxsubl2
+                        maxsubr2 = np.max(
+                            lg2.margin_vals['right'][ss2.colspan[:-1]] +
+                            lg2.margin_vals['rightcb'][ss2.colspan[:-1]])
+                        if maxsubr2 > maxsubr:
+                            maxsubr = maxsubr2
+            for i in ss1.colspan[1:]:
+                lg1.edit_margin_min('left', maxsubl, col=i)
+            for i in ss1.colspan[:-1]:
+                lg1.edit_margin_min('right', maxsubr, col=i)
+
+        # interior rows:
+        nc = len(ss1.rowspan)
+        if nc > 1:
+            maxsubt = np.max(
+                lg1.margin_vals['top'][ss1.rowspan[1:]] +
+                lg1.margin_vals['topcb'][ss1.rowspan[1:]]
+            )
+            maxsubb = np.max(
+                lg1.margin_vals['bottom'][ss1.rowspan[:-1]] +
+                lg1.margin_vals['bottomcb'][ss1.rowspan[:-1]]
+            )
+
+            for ax2 in axs:
+                ss2 = ax2.get_subplotspec()
+                lg2 = ss2.get_gridspec()._layoutgrid
+                if lg2 is not None:
+                    nc = len(ss2.rowspan)
+                    if nc > 1:
+                        maxsubt2 = np.max(
+                            lg2.margin_vals['top'][ss2.rowspan[1:]] +
+                            lg2.margin_vals['topcb'][ss2.rowspan[1:]]
+                        )
+                        if maxsubt2 > maxsubt:
+                            maxsubt = maxsubt2
+                        maxsubb2 = np.max(
+                            lg2.margin_vals['bottom'][ss2.rowspan[:-1]] +
+                            lg2.margin_vals['bottomcb'][ss2.rowspan[:-1]]
+                        )
+                        if maxsubb2 > maxsubb:
+                            maxsubb = maxsubb2
+            for i in ss1.rowspan[1:]:
+                lg1.edit_margin_min('top', maxsubt, col=i)
+            for i in ss1.rowspan[:-1]:
+                lg1.edit_margin_min('bottom', maxsubb, col=i)
+
+
+def _get_cb_parent_spans(cbax):
+    """
+    Figure out which subplotspecs this colorbar belongs to
+    """
+    rowspan = range(100, -100, -100)
```

Comment:
sys.maxsize should keep you safe in case anyone ever has more than 100x100 axes...
(also step can stay None, I guess)